executor/write.go

```diff
@@ -907,17 +941,27 @@ func getKeysNeedCheck(ctx sessionctx.Context, t table.Table, rows [][]types.Datu
 	}
 	rowWithKeys := make([][]keyWithDupError, 0, len(rows))
 
-	// get recordIDs
-	recordIDs, err := getRecordIDs(ctx, t, rows)
+	// Get the new handles.
+	handles, err := genNewHandles(ctx, t, rows)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
 	for i, row := range rows {
 		keysWithErr := make([]keyWithDupError, 0, nUnique+1)
+		newRowValue, err := encodeNewRow(ctx, t, row)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
 		// append record keys and errors
 		if t.Meta().PKIsHandle {
-			keysWithErr = append(keysWithErr, keyWithDupError{t.RecordKey(recordIDs[i]), kv.ErrKeyExists.FastGen("Duplicate entry '%d' for key 'PRIMARY'", recordIDs[i])})
+			keysWithErr = append(keysWithErr, keyWithDupError{
+				true,
+				t.RecordKey(handles[i]),
+				kv.ErrKeyExists.FastGen("Duplicate entry '%d' for key 'PRIMARY'", handles[i]),
+				handles[i],
+				newRowValue,
+			})
 		}
```

Comment:
How about `else ` here?
For `PKIsHandle` table, we only need to BatchGet handles.